spacy_llm/ty.py

```diff
@@ -20,3 +23,138 @@ def parse_responses(
         self, docs: Iterable[Doc], responses: Iterable[_Response]
     ) -> Iterable[Doc]:
         ...
+
+
+def _do_args_match(out_arg: Iterable, in_arg: Iterable) -> bool:
+    """Compares argument type of Iterables for compatibility.
+    in_arg (Iterable): Input argument.
+    out_arg (Iterable): Output argument.
+    RETURNS (bool): True if type variables are of the same length and if type variables in out_arg are a subclass
+        of (or the same class as) the type variables in in_arg.
+    """
+    assert hasattr(out_arg, "__args__") and hasattr(in_arg, "__args__")
+    # Replace Any with object to make issubclass() check work.
+    out_type_vars = [arg if arg != Any else object for arg in out_arg.__args__]
+    in_type_vars = [arg if arg != Any else object for arg in in_arg.__args__]
+
+    if len(out_type_vars) != len(in_type_vars):
+        return False
+
+    return all(
+        [
+            issubclass(out_tv, in_tv) or issubclass(in_tv, out_tv)
+            for out_tv, in_tv in zip(out_type_vars, in_type_vars)
+        ]
+    )
+
+
+def _extract_backend_call_signature(backend: PromptExecutor) -> Dict[str, Any]:
+    """Extract call signature from backend object.
+    backend (PromptExecutor): Backend object to extract call signature from.
+    RETURNS (Dict[str, Any]): Type per argument name.
+    """
+    if hasattr(backend, "__call__"):
+        signature = typing.get_type_hints(backend.__call__)
+        # We want to check whether the supplied backend is a Generic type, because that requires a different approach to
+        # extracting the proper signature types. There doesn't seem to be an official way to do this for Python 3.6, so
+        # for compatibility we check for __orig_class__.
+        if not hasattr(backend, "__orig_class__"):
+            return signature
+
+        # If __call__() uses generic types, typing.get_type_hints() only yields generic aliases, which we aren't
+        # interested in. In this case we have to do some type trickery to fetch the actual types.
+        # We also prohibit the usage of any generic objects apart from integration.Backend, as this would make type
+        # checking more complicated (and there probably isn't much demand for this anyway).
+        if not isinstance(backend, integration.Backend):
+            raise ValueError(
+                "Generics in backend objects are currently not supported when not using spacy_llm."
+                "backends.integration.Backend."
+            )
+        return {
+            "input": Iterable[backend.__orig_class__.__args__[1]],  # type: ignore[name-defined]
+            "return": Iterable[backend.__orig_class__.__args__[2]],  # type: ignore[name-defined]
+        }
```

Comment:
This is new and should be reviewed. The rest of the type checking routines in `ty.py` were just moved.